wlauto/workloads/googlephotos/__init__.py

```diff
@@ -0,0 +1,104 @@
+#    Copyright 2014-2016 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+
+from wlauto import AndroidUiAutoBenchmark, Parameter, File
+from wlauto.exceptions import ValidationError
+from wlauto.utils.types import list_of_strings
+
+
+class Googlephotos(AndroidUiAutoBenchmark):
+
+    name = 'googlephotos'
+    package = 'com.google.android.apps.photos'
+    version = '1.21.0.123444480'  # apk version
+    activity = 'com.google.android.apps.photos.home.HomeActivity'
+    view = [package + '/com.google.android.apps.consumerphotoeditor.fragments.ConsumerPhotoEditorActivity',
+            package + '/com.google.android.apps.photos.home.HomeActivity',
+            package + '/com.google.android.apps.photos.localmedia.ui.LocalPhotosActivity',
+            package + '/com.google.android.apps.photos.onboarding.AccountPickerActivity',
+            package + '/com.google.android.apps.photos.onboarding.IntroActivity']
+    description = """
+    A workload to perform standard productivity tasks with Google Photos. The workload carries out
+    various tasks, such as browsing images, performing zooms, and post-processing the image.
+
+    Test description:
+    1. Four images are copied to the device
+    2. The application is started in offline access mode
+    3. Gestures are performed to swipe between images and pinch zoom in and out of the selected
+       image
+    4. The colour of a selected image is edited by selecting the colour menu, incrementing the
+       colour, resetting the colour and decrementing the colour using the seek bar.
+    5. A crop test is performed on a selected image.  UiAutomator does not allow the selection of
+       the crop markers so the image is tilted positively, reset and then negatively to get a
+       similar cropping effect.
+    6. A rotate test is performed on a selected image, rotating anticlockwise 90 degrees, 180
+       degrees and 270 degrees.
+    """
+
+    default_test_images = [
+        'uxperf_1200x1600.png', 'uxperf_1600x1200.jpg',
+        'uxperf_2448x3264.png', 'uxperf_3264x2448.jpg',
+    ]
+
+    parameters = [
+        Parameter('markers_enabled', kind=bool, default=True,
+                  description="""
+                  If ``True``, UX_PERF action markers will be emitted to logcat during
+                  the test run.
+                  """),
+        Parameter('test_images', kind=list_of_strings, default=default_test_images,
```

Comment:
This list must contain exactly four unique images. This should be documented by adding a validation constraint to this parameter:

      constraint=lambda x: len(unique(x)) == 4

The constraint will be automatically tested when setting the parameter, removing the need for explicit check inside `validate` and it will also appear in the documentation for the workload.

(note: `unique` may be imported from `wlauto.utils.misc`; it's used here istead of `set` for documentation purposes).